### PR TITLE
Add CKComponentBoundsAnimation to CKCollectionViewTransactionalDataSource

### PR DESCRIPTION
**Pull request by conradev**
In `CKCollectionViewDataSource`, synchronous cell updates do not trigger a collection view reload and instead mount the new component layout in place and animate any size changes. I made this the default behavior for both asynchronous and synchronous updates to `CKCollectionViewTransactionalDataSource`, where the content of the changeset is only cell updates.

Calling `-[UICollectionView reloadItemsAtIndexPaths:]` causes the component to be unmounted and remounted. This is fine in some cases, but it has adverse side effects. It causes the cell to flash to its new state without animation and it also disrupts the first responder.

### CODE REVIEW

**Comment by conradev** [comment]
Rebased against master!

**Comment by conradev** [comment]
and fixed tests

**Comment by facebook-github-bot-3** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @MDSilber as a potential reviewer. Could you take a look please or cc someone with more context?

File: ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm [hunk 1]
```diff
@@ -84,14 +85,57 @@ - (void)transactionalComponentDataSource:(CKTransactionalComponentDataSource *)d
...
 {
-  [_collectionView performBatchUpdates:^{
-    applyChangesToCollectionView(changes, _collectionView);
-    // Detach all the component layouts for items being deleted
-    [self _detachComponentLayoutForRemovedItemsAtIndexPaths:[changes removedIndexPaths]
-                                                    inState:previousState];
-    // Update current state
-    _currentState = [_componentDataSource state];
-  } completion:NULL];
+  BOOL hasMutations = (changes.removedIndexPaths.count ||
```
Comment (ocrickard) [inline ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm]: const

File: ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm [hunk 2]
```diff
@@ -84,14 +85,57 @@ - (void)transactionalComponentDataSource:(CKTransactionalComponentDataSource *)d
...
+                      changes.insertedIndexPaths.count ||
+                      changes.movedIndexPaths.count ||
+                      changes.insertedSections.count ||
+                      changes.removedSections.count);
+  BOOL hasUpdates = !!changes.updatedIndexPaths.count;
```
Comment (ocrickard) [inline ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm]: const

File: ComponentKit/DataSources/CKComponentBoundsAnimation+UICollectionView.mm [hunk 3]
```diff
@@ -103,6 +103,9 @@ - (void)applyBoundsAnimationToCollectionView:(const CKComponentBoundsAnimation &
   if (animation.duration == 0) {
     return;
   }
+  if (!_collectionView.window) {
```
Comment (ocrickard) [inline ComponentKit/DataSources/CKComponentBoundsAnimation+UICollectionView.mm]: Needs a comment.

**Comment by ocrickard** [comment]
needs to be rebased. defer to @eczarny

**Comment by eczarny** [comment]
@conradev mind resolving conflicts? More than happy to finish up review and get this merged ASAP.

**Comment by conradev** [comment]
Awesome! Let me know if you have any feedback, excited to get this finally merged 😀

**Comment by eczarny** [comment]
Thanks! I'll review this today.

File: ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm [hunk 4]
```diff
@@ -93,14 +94,56 @@ - (void)transactionalComponentDataSource:(CKTransactionalComponentDataSource *)d
...
+    
+    CKComponentBoundsAnimationApply(boundsAnimation, ^{
+      for (NSIndexPath *indexPath in changes.updatedIndexPaths) {
+        CKTransactionalComponentDataSourceItem *item = [state objectAtIndexPath:indexPath];
+        CKCollectionViewDataSourceCell *cell = (CKCollectionViewDataSourceCell *)[_collectionView cellForItemAtIndexPath:indexPath];
+        if (cell) {
+          attachToCell(cell, item, _attachController);
+        }
+      }
+    }, nil);
```
Comment (eczarny) [inline ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm]: There's quite a bit going on here. Comments explaining what we're doing would be really helpful.

File: ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm [hunk 5]
```diff
@@ -159,8 +202,7 @@ - (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cell
 {
   CKCollectionViewDataSourceCell *cell = [_collectionView dequeueReusableCellWithReuseIdentifier:kReuseIdentifier forIndexPath:indexPath];
   CKTransactionalComponentDataSourceItem *item = [_currentState objectAtIndexPath:indexPath];
-  attachToCell(cell, item, _attachController);
-  [_cellToItemMap setObject:item forKey:cell];
```
Comment (eczarny) [inline ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm]: Is this safe to drop? Wouldn't this lead to broken appearance announcements?

File: ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm [hunk 6]
```diff
@@ -93,14 +94,56 @@ - (void)transactionalComponentDataSource:(CKTransactionalComponentDataSource *)d
...
+    CKComponentBoundsAnimationApply(boundsAnimation, ^{
+      for (NSIndexPath *indexPath in changes.updatedIndexPaths) {
+        CKTransactionalComponentDataSourceItem *item = [state objectAtIndexPath:indexPath];
+        CKCollectionViewDataSourceCell *cell = (CKCollectionViewDataSourceCell *)[_collectionView cellForItemAtIndexPath:indexPath];
+        if (cell) {
+          attachToCell(cell, item, _attachController);
+        }
+      }
+    }, nil);
+  } else if (hasMutations) {
```
Comment (eczarny) [inline ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm]: Does this check change the behavior at all? It looks like prior to this pull request the logic below was always performed. Does it make sense to check `hasMutations` or can we just replace the `else if (...` with an `else`?

File: ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm [hunk 7]
```diff
@@ -93,14 +94,56 @@ - (void)transactionalComponentDataSource:(CKTransactionalComponentDataSource *)d
...
+      if (boundsAnimation.duration)
+        break;
+    }
+    
+    if (boundsAnimation.duration) {
+      id boundsAnimationContext = CKComponentBoundsAnimationPrepareForCollectionViewBatchUpdates(_collectionView);
+      [UIView performWithoutAnimation:^{
+        [_collectionView performBatchUpdates:^{
+          _currentState = state;
+        } completion:^(BOOL finished) {}];
```
Comment (eczarny) [inline ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm]: Since this is called twice could we put it in a block and just execute the block here or below, on line 123? Might be a bit more obvious. It's also a pattern I've seen used elsewhere.

File: ComponentKit.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -3404,10 +3404,6 @@
 		B342DC471AC23E8200ACAC53 /* Debug */ = {
 			isa = XCBuildConfiguration;
 			buildSettings = {
-				FRAMEWORK_SEARCH_PATHS = (
-					"$(inherited)",
-					"$(PROJECT_DIR)/Carthage/Build/iOS",
-				);
```
Comment (conradev) [inline ComponentKit.xcodeproj/project.pbxproj]: Yes – they are in a separate commit to fix a build warning. Can move to another PR if that makes more sense.

File: ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm (hunk 6 above)
Comment (conradev) [inline ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm]: It does change the behavior. This change means that empty changesets do nothing when applied, which I think is a good change to make.

File: ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm (hunk 5 above)
Comment (conradev) [inline ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm]: Oops! Good catch. I think this might've been an error during rebasing. It turns out that `_cellToItemMap` was being neglected in other circumstances as well, so I will go ahead and fix that as well.

**Comment by conradev** [comment]
Went ahead and made those changes, let me know if you have further feedback! 😄

**Comment by eczarny** [comment]
Thanks for the quick turnaround! I'll review your changes first thing in the morning.

File: ComponentKit.xcodeproj/project.pbxproj (hunk 8 above)
Comment (eczarny) [inline ComponentKit.xcodeproj/project.pbxproj]: No need to create a separate pull request. Thanks for the cleanup!

File: ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm [hunk 9]
```diff
@@ -93,14 +95,65 @@ - (void)transactionalComponentDataSource:(CKTransactionalComponentDataSource *)d
...
-                                                    inState:previousState];
-    // Update current state
-    _currentState = [_componentDataSource state];
-  } completion:NULL];
+  const BOOL changesIncludeNonUpdates = (changes.removedIndexPaths.count ||
+                                         changes.insertedIndexPaths.count ||
+                                         changes.movedIndexPaths.count ||
+                                         changes.insertedSections.count ||
+                                         changes.removedSections.count);
+  const BOOL changesIncludeOnlyUpdates = (!!changes.updatedIndexPaths.count && !changesIncludeNonUpdates);
```
Comment (eczarny) [inline ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm]: `!!changes.updatedIndexPaths.count`? Is that correct? If you intend to check that updated index paths do exist `changes.updatedIndexPaths.count > 0` would be preferable.

**Comment by eczarny** [comment]
@facebook-github-bot import

File: ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm [hunk 10]
```diff
@@ -93,14 +95,65 @@ - (void)transactionalComponentDataSource:(CKTransactionalComponentDataSource *)d
...
+    // We are not able to animate the updates individually, so we pick the
+    // first bounds animation with a non-zero duration.
+    CKComponentBoundsAnimation boundsAnimation = {};
+    for (NSIndexPath *indexPath in changes.updatedIndexPaths) {
+      boundsAnimation = [[state objectAtIndexPath:indexPath] boundsAnimation];
+      if (boundsAnimation.duration)
+        break;
+    }
+    
+    void (^updateState)(CKTransactionalComponentDataSourceState *) = ^(CKTransactionalComponentDataSourceState *state) {
```
Comment (eczarny) [inline ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm]: `state` should be renamed as it shadows the local variable declared on line 105.

**Comment by conradev** [comment]
Made those changes!

**Comment by eczarny** [comment]
@facebook-github-bot import

**Comment by eczarny** [comment]
I'm back. I have a few more things to finish on my side but I hope to merge this soon.

**Comment by razvanpaul** [comment]
I'm trying to integrate the CKCollectionViewTransactionalDataSource, but I cannot implement the following methods of the UICollectionViewDelegate because apparently they don't exist anymore:
`/**
    Sends -componentTreeWillAppear to all CKComponentControllers for the given cell.
    If needed, call this from -collectionView:willDisplayCell:forItemAtIndexPath:
 */
- (void)announceWillDisplayCell:(UICollectionViewCell *)cell;

/**
    Sends -componentTreeDidDisappear to all CKComponentControllers for the given cell.
    If needed, call this from -collectionView:didEndDisplayingCell:forItemAtIndexPath:
 */
- (void)announceDidEndDisplayingCell:(UICollectionViewCell *)cell;
`

What is the current way to do this? Or it is not necessary right now and there is another way to do it?

**Comment by conradev** [comment]
Let's shoot to keep this discussion in one place (in #726)
